---
title: "m6A_site_motif_SNP"
author: "DP"
date: "2023-10-10"
output: html_document
---
## ggplot color 
https://r-graph-gallery.com/ggplot2-color.html
https://derekogle.com/NCGraphing/resources/colors
## Load packages
```{r}
library("reshape2")
library("data.table")
library("dplyr")
library("stringr")
library("seqinr") 
library("tidyverse")

# packages for plot 
library("ggplot2")
library("ggpubr")
library("pheatmap")
library("gridExtra")
library("grid")
library("cowplot")
library("ggseqlogo")


## Default renaming
rename_default <- function(names){
  return(unlist(strsplit(names, split = "[|]"))[6])
}

rename_default_5utr <- function(names){
  pre = unlist(strsplit(names, split = "[|]"))[8] 
  return(unlist(strsplit(pre, split = "[-]"))[2] )
}

rename_default_cds <- function(names){
  pre = unlist(strsplit(names, split = "[|]"))[9] 
  return(unlist(strsplit(pre, split = "[-]"))[2] )

}

rename_default_3utr <- function(names){
  return(unlist(strsplit(names, split = "[|]"))[7] )

}

```

### Load data 
```{r}
## Load filtered m6A sites datasets
## This dataframe were generated from Mouse_Fig3_ASM.Rmd (Line 324)
step6_prob85 = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/processed_data/step6_prob85.csv", sep = ",")
step6_prob85$Gene = paste0(step6_prob85$gene, "_", step6_prob85$transcript_position)

## Load determined ASM datasets
## This dataframe were generated from Mouse_Fig3_ASM.Rmd (Line 439)
mouse_ASM = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/processed_data/mouse_ASM.csv", sep = ",")
mouse_ASM_mod <- step6_prob85[step6_prob85$Gene %in% mouse_ASM$Gene, ]

```

## mouse reference sequences
```{r}
mouse_transcript_masked = readLines("/Users/dayeapark/Desktop/Script_upload/datasets/reference/variant_masked_mouse_transcriptome.fa") 
mouse_ALT_unmasked = readLines("/Users/dayeapark/Desktop/Script_upload/datasets/reference/ALT_variant_masked_mouse_transcriptome.fa")
mouse_REF_unmasked = readLines("/Users/dayeapark/Desktop/Script_upload/datasets/reference/REF_variant_masked_mouse_transcriptome.fa")

line_to_table <-function (df) {
  gene_lines <- seq(1, length(df)-1, by=2)
  sequence_lines <- seq(2, length(df)-1, by=2)
  
  mouse_transcript_df <- data.frame( transcript_id = df[gene_lines], sequence = df[sequence_lines] )
  mouse_transcript_df$gene = sapply(mouse_transcript_df$transcript_id, rename_default)
  
  return(mouse_transcript_df)
}

## reference sequences 
masked_transcript_seq = line_to_table(mouse_transcript_masked)
unmasked_ALT_seq = line_to_table(mouse_ALT_unmasked)
unmasked_REF_seq = line_to_table(mouse_REF_unmasked)


```

#################################################################################
### Functions for SNPs on DRACH motif
```{r}
## SNP search on motif 
SNP_mark <- function(df, pos_num, mark) {
  df$transcript_pos = df$transcript_position + pos_num
  site_mark <- merge(mouse_vcf_snp, df, by = c("gene", "transcript_pos"))
  
  if (nrow(site_mark) > 0 ) { site_mark$SNP_pos = mark}
  
  return(site_mark)
}


## ggmotif, https://omarwagih.github.io/ggseqlogo/
## merge sequence from reference files. 
add_entire_sequence <- function (df)  { 
for (i in 1:nrow(df)) {
  df$masked_seq[i] = masked_transcript_seq[masked_transcript_seq$gene == df$gene[i],]$sequence
  df$ALT_seq[i] = unmasked_ALT_seq[unmasked_ALT_seq$gene == df$gene[i],]$sequence
  df$REF_seq[i] = unmasked_REF_seq[unmasked_REF_seq$gene == df$gene[i],]$sequence
}
return(df)
}

## kmer included data frame 
motif_seq_search <-function (df, adding_num_start, adding_num_end, kmer_num) {
    df_motif = add_entire_sequence(df)
    df_motif$start_pos = df_motif$transcript_position + adding_num_start
    df_motif$end_pos = df_motif$transcript_position + adding_num_end
      
    df_motif$masked_motif_seq = substr(df_motif$masked_seq, df_motif$start_pos, df_motif$end_pos)
    df_motif$ALT_motif_seq = substr(df_motif$ALT_seq, df_motif$start_pos, df_motif$end_pos)
    df_motif$REF_motif_seq = substr(df_motif$REF_seq, df_motif$start_pos, df_motif$end_pos)
      
    nuc_motif_search = df_motif[,c("gene", "transcript_position", "Gene", 
                            "rep1_alt_mod_ratio", "rep1_ref_mod_ratio", "rep2_alt_mod_ratio", "rep2_ref_mod_ratio", 
                            "masked_motif_seq", "ALT_motif_seq", "REF_motif_seq")]
    
    final_df <- nuc_motif_search[nchar(nuc_motif_search$masked_motif_seq ) == kmer_num, ]
    final_df$alt_mod_ratio_mean = (final_df$rep1_alt_mod_ratio + final_df$rep2_alt_mod_ratio)/2
    final_df$ref_mod_ratio_mean = (final_df$rep1_ref_mod_ratio + final_df$rep2_ref_mod_ratio)/2

  return(final_df) 
}


## SNP analysis functions (SNP pos)

Search_for_SNP_pos_mouse <- function (df, ASM_list) { 
    
    add_line <- data.frame(gene = NA, transcript_position= NA, Gene= NA, 
                               rep1_alt_mod_ratio = NA, rep1_ref_mod_ratio = NA, rep2_alt_mod_ratio = NA, rep2_ref_mod_ratio= NA,
                               masked_motif_seq = NA, ALT_motif_seq = NA, REF_motif_seq = NA, 
                               alt_mod_ratio_mean = NA, ref_mod_ratio_mean = NA, SNP_pos = NA, SNP_nuc_highm6A = NA, SNP_nuc_lowm6A = NA)
    for (i in 1:nrow(df)) { 
      if (str_count(df$masked_motif_seq[i], pattern ="N") == 0 ){
        df$SNP_pos[i] = "none" 
        df$SNP_nuc_highm6A[i] = "none" 
        df$SNP_nuc_lowm6A[i] = "none" 
      } else if (str_count(df$masked_motif_seq[i], pattern ="N") == 1 ){
        for (pos_num in 1:nchar(df$masked_motif_seq[i])) { 
          if(substr(df$masked_motif_seq[i], pos_num, pos_num) == "N" ) {
            df$SNP_pos[i] = pos_num 
            
            if (df$alt_mod_ratio_mean[i] > df$ref_mod_ratio_mean[i]) { 
            df$SNP_nuc_highm6A[i] = substr(df$ALT_motif_seq[i], pos_num, pos_num)
            df$SNP_nuc_lowm6A[i] = substr(df$REF_motif_seq[i], pos_num, pos_num)
            } else if (df$alt_mod_ratio_mean[i] < df$ref_mod_ratio_mean[i]) { 
            df$SNP_nuc_highm6A[i] = substr(df$REF_motif_seq[i], pos_num, pos_num)
            df$SNP_nuc_lowm6A[i] = substr(df$ALT_motif_seq[i], pos_num, pos_num)}
          }}
          
      } else if (str_count(df$masked_motif_seq[i], pattern ="N") > 1 ){
        add_line_new = df %>% slice(rep(i, each = str_count(df$masked_motif_seq[i], pattern ="N")))  
        list_pos = data.frame(str_locate_all(pattern ='N', df$masked_motif_seq[i]))
        for (snp_n in 1:nrow(add_line_new)) {
          add_line_new$SNP_pos[snp_n] = list_pos[snp_n,1]

            if (add_line_new$alt_mod_ratio_mean[snp_n] > add_line_new$ref_mod_ratio_mean[snp_n]) { 
            add_line_new$SNP_nuc_highm6A[snp_n] = substr(add_line_new$ALT_motif_seq[snp_n], list_pos[snp_n,1], list_pos[snp_n,1] )
            add_line_new$SNP_nuc_lowm6A[snp_n] = substr(add_line_new$REF_motif_seq[snp_n], list_pos[snp_n,1], list_pos[snp_n,1] )
            } else if (add_line_new$alt_mod_ratio_mean[snp_n] < add_line_new$ref_mod_ratio_mean[snp_n]) { 
            add_line_new$SNP_nuc_highm6A[snp_n] = substr(add_line_new$REF_motif_seq[snp_n], list_pos[snp_n,1], list_pos[snp_n,1] )
            add_line_new$SNP_nuc_lowm6A[snp_n] = substr(add_line_new$ALT_motif_seq[snp_n], list_pos[snp_n,1], list_pos[snp_n,1] )
            }} 
        add_line = rbind(add_line, add_line_new) }
    }
    
    final_df = na.omit (rbind(add_line, df[!(df$Gene %in% add_line$Gene), ]) ) 
    final_df$ASM_type = "none"
    final_df[final_df$Gene %in% mouse_ASM$Gene,]$ASM_type = "ASM"
      
  return(final_df)
}


```

################################ Group 1 ASM ###################################
## PLOT: Group 1 N position (Figure 4A)
### motif preferences were observed by finding 5mer sequences 
```{r}
mouse_prob85_5mer = motif_seq_search( step6_prob85, (-1), 3, 5 ) 

ASM_5mer = mouse_prob85_5mer[mouse_prob85_5mer$Gene %in% mouse_ASM$Gene,]

ASM_5mer$ASM_group = "group2"
ASM_5mer[str_count(ASM_5mer$masked_motif_seq, "N") > 0, ]$ASM_group = "group1"

#ggplot_logo(ASM_5mer[ASM_5mer$ASM_group == "group1", ], "masked_motif_seq", "ASM gourp1:5-mer masked seq")

Fig4A_df = data.frame ( site = c("D", "R", "A", "C", "H"), 
                             count = c(4, 1, 0, 0, 2), 
                             level = c(1,2,3,4,5))

Fig4A_df <- ggplot(data = Fig4A_df, aes(x=reorder(site, level), y = count)) +
              geom_col(width = 0.7, fill = "darkblue") +
              theme_classic() +
              theme(legend.position= "none")

Fig4A_df
```
### Group1 modification ratio (Figure 4B)
```{r}
mESC_all_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep1_all.csv.gz", header = TRUE, sep = ",")
mESC_all_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep2_all.csv.gz", header = TRUE, sep = ",")
mESC_all_rep1_max$gene = sapply(mESC_all_rep1_max$transcript_id, rename_default ) 
mESC_all_rep2_max$gene = sapply(mESC_all_rep2_max$transcript_id, rename_default ) 
mESC_all_rep1_max$Gene = paste0(mESC_all_rep1_max$gene, "_", mESC_all_rep1_max$transcript_position)
mESC_all_rep2_max$Gene = paste0(mESC_all_rep2_max$gene, "_", mESC_all_rep2_max$transcript_position)
mESC_all_rep1_max$rep = "rep1"
mESC_all_rep2_max$rep = "rep2"

all_sites_mouse = rbind(mESC_all_rep1_max[,c(7,2,9,3:6,8)], mESC_all_rep2_max[,c(7,2,9,3:6,8)])

all_sites_mouse_prob85 = all_sites_mouse[all_sites_mouse$probability_modified > 0.85 ,]
kmer_prob85 <- all_sites_mouse_prob85[all_sites_mouse_prob85$Gene %in% step6_prob85$Gene, c(3,6,7)]
kmer_prob85_mean <- kmer_prob85 %>% 
                      group_by(kmer) %>% 
                      summarise(mean_prob = mean(mod_ratio),
                                freq = n())

## Figure 4B (top)
ggplot(kmer_prob85_mean, aes(x=reorder(kmer, -freq), y=freq)) + 
  geom_bar(stat = "identity")  +
 geom_text(aes(y = freq, label = freq)) +
  theme_classic() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


## Group1 matrix
group1_gene = ASM_5mer[ASM_5mer$ASM_group == "group1",]$Gene
group1_table = matrix(0, nrow  =  length(group1_gene), 
                      ncol = nrow(kmer_prob85_mean))

colnames(group1_table) = kmer_prob85_mean$kmer
rownames(group1_table) = group1_gene
alt = ASM_5mer[ASM_5mer$Gene %in% group1_gene, c(3,9,11)]
ref = ASM_5mer[ASM_5mer$Gene %in% group1_gene, c(3,10,12)]
colnames(alt) = colnames(ref) = c("Gene", "kmer", "modratio")
group1_mod_ratio = rbind(alt, ref)

for (i in 1:nrow(group1_mod_ratio)) {
  group1_table[ group1_mod_ratio$Gene[i], group1_mod_ratio$kmer[i] ] <- group1_mod_ratio$modratio[i]
}


all_mod_ratio_pre = data.frame(kmer_prob85_mean[,c(1,2)])
all_mod_ratio_pre$Gene = "all"
all_mod_ratio = all_mod_ratio_pre[,c(3,1,2)]
colnames(all_mod_ratio) <- colnames(group1_mod_ratio)

include_all_ASM_moratio = data.table(rbind(all_mod_ratio, group1_mod_ratio))


merge_all_fre_mod <- merge(include_all_ASM_moratio, kmer_prob85_mean[,c(1,3)], all = TRUE)

## Figure 4B (bottom)
ggplot(merge_all_fre_mod, aes(x = reorder(kmer, -freq) , y = Gene, fill = modratio)) + 
  geom_tile() +
  scale_fill_gradientn(colours=c("white","red"), limits=c(0.2,1)) + 
  #scale_fill_gradient(low="white", high="red") + 
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```


################################ Group 2 ASM ###################################
## SNPs on  +/-1 or +/-2 DRACH motif 
We observed SNPs near DRAFH motif shows specific nucleotide (-1 site >> "C") in cluster 2 ASM genes. 
Therefore, I will search the nucleotides of + or -1 DRACH motif and compare modification. 

### 9-mers: SNP Analysis 
```{r}
## all prob85: 9-mer 
mouse_prob85_9mer = motif_seq_search( step6_prob85, (-3), 5, 9 ) # D-2 site to H+2 site >> 9 mer

## including non-SNP nucleotides (9mer)
mouse_prob85_9mer_snp_pos_ASM = Search_for_SNP_pos_mouse(mouse_prob85_9mer, mouse_ASM$Gene)

## ASM sites
ASM_seq_SNP_included= mouse_prob85_9mer_snp_pos_ASM[mouse_prob85_9mer_snp_pos_ASM$Gene %in% mouse_ASM$Gene, ]

## Group 1 ASM which contains SNPs in DRACH motif
Group1_ASM = ASM_seq_SNP_included[ASM_seq_SNP_included$SNP_pos %in% c(3,4,5,6,7),]$Gene

## Group 2 ASM which does not contains SNPs in DRACH motif 
Group2_ASM = mouse_ASM_mod[!(mouse_ASM_mod$Gene %in% Group1_ASM), ]$Gene

```

### PLOT: SNPs located in adjacent motif - Group 2 ASM (Figure 4C)
```{r}
ggplot_logo <- function (df, selec_column, title) {
  ggplot() + 
  #geom_logo( df %>% select (selec_column) , method = 'prob') + 
  geom_logo( df %>% select (selec_column) , method = 'bit') + 
  theme_logo() +
  ggtitle(title) + 
  theme_classic() +
  theme(plot.title = element_text(size = 10),
        legend.position = "none")
  

}

select_mout_ASM <- ASM_seq_SNP_included[,c("Gene", "masked_motif_seq")]
select_mout_ASM$T_to_U  = gsub("T","U", select_mout_ASM$masked_motif_seq) 
ggplot_logo(ASM_seq_SNP_included, "masked_motif_seq", "sigASM:9-mer masked seq")
ggplot_logo(ASM_seq_SNP_included[ASM_seq_SNP_included$ASM_type == "ASM" & 
                                ASM_seq_SNP_included$Gene %in% Group2_ASM , ], 
            "masked_motif_seq", "sigASM:9-mer masked seq")

```



## PLOT: Group 2 ASM Motif sequence preference (Figure 4E)
```{r}
## plot function 
motif_freq_plot <- function (df, title_pick) {
  df_freq = data.frame(table(df$masked_motif_seq))
  
  ggplot(data = df_freq, aes(x=reorder(Var1, Freq), y = Freq)) +
        geom_col(width = 0.7, fill = "darkgrey") +
        theme_classic() +
        theme(legend.position= "none") + 
        ggtitle(title_pick)
}

## Group 2 motif preference 
motif_freq_plot(mouse_prob85_5mer, "entire motif sequence frequency")   ## for all 
group2_5mer <- ASM_5mer[ASM_5mer$ASM_group == "group2", ]
group2_moditf <- motif_freq_plot(group2_5mer, "Group 2 ASM motif sequence frequency") 
group2_moditf
```
