---
title: "ASM determination and analysis"
author: "DP"
date: "2023-09-07"
output: html_document
---

## Load packages
```{r}
library("reshape2")
library("data.table")
library("dplyr")
library("stringr")
library("tidyverse")
library("palmerpenguins")
library("data.table")

# packages for plot 
library("ggplot2")
library("ggpubr")
library("ggVennDiagram")
library("pheatmap")
library("gridExtra")


```
## Load data (m6anet output)
```{r}
# Reference: maternal / Alternative: paternal reads 
## mESC_rep2_max
## Max n_read: all (20k), alt (10k), ref (10k), und (12k)
mESC_all_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep1_all.csv.gz", header = TRUE, sep = ",")
mESC_alt_ALT_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_mESC_rep1_alt_ALT.csv.gz", header = TRUE, sep = ",")
mESC_ref_REF_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_mESC_rep1_ref_REF.csv.gz", header = TRUE, sep = ",")
mESC_und_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep1_und.csv.gz", header = TRUE, sep = ",")

## mESC_rep2_max
## Max n_read: all (15k), alt (10k), ref (10k), und (12k)
mESC_all_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep2_all.csv.gz", header = TRUE, sep = ",")
mESC_alt_ALT_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_mESC_rep2_alt_ALT.csv.gz", header = TRUE, sep = ",")
mESC_ref_REF_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_mESC_rep2_ref_REF.csv.gz", header = TRUE, sep = ",")
mESC_und_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep2_und.csv.gz", header = TRUE, sep = ",")


## mettl3
mettl3_all_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mettl3_all.csv.gz", header = TRUE, sep = ",")
mettl3_alt_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mettl3_alt.csv.gz", header = TRUE, sep = ",")
mettl3_ref_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mettl3_ref.csv.gz", header = TRUE, sep = ",")
mettl3_und_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mettl3_und.csv.gz", header = TRUE, sep = ",")

## maksed
mESC_alt_masked_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep1_alt.csv.gz", header = TRUE, sep = ",")
mESC_ref_masked_rep1_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep1_ref.csv.gz", header = TRUE, sep = ",")
mESC_alt_masked_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep2_alt.csv.gz", header = TRUE, sep = ",")
mESC_ref_masked_rep2_max = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/m6Anet_output/m6Anet_max_mESC_rep2_ref.csv.gz", header = TRUE, sep = ",")

```
## reframe m6Anet output for further analysis 
```{r}
reframe_m6Anet <- function (df) {
  df$gene = sapply(df$transcript_id, rename_default)
  df_select = df[,c(7,2:6)]
  return(df_select)
}

make_merge_file <- function(m6Anet_all, m6Anet_alt, m6Anet_ref, m6Anet_und) {
  
  df_all = reframe_m6Anet(m6Anet_all)
  df_alt = reframe_m6Anet(m6Anet_alt)
  df_ref = reframe_m6Anet(m6Anet_ref)
  df_und = reframe_m6Anet(m6Anet_und)
  
  column_select = c("gene", "transcript_position", "n_reads", "probability_modified", "mod_ratio"  )
  df_all_new = df_all[,column_select]
  df_alt_new = df_alt[,column_select]
  df_ref_new = df_ref[,column_select]
  df_und_new = df_und[,column_select]
  merge_df1 = merge(df_all_new, df_alt_new, by = c("gene", "transcript_position"), all.x = TRUE)
  merge_df2 = merge(merge_df1, df_ref_new, by = c("gene", "transcript_position"), all.x = TRUE)
  new_merge_df = merge(merge_df2, df_und_new, by = c("gene", "transcript_position"), all.x = TRUE)
  
  colnames( new_merge_df) = c("gene", "transcript_position",
                              "all_n_reads", "all_prob_mod", "all_mod_ratio",
                              "alt_n_reads", "alt_prob_mod", "alt_mod_ratio",
                              "ref_n_reads", "ref_prob_mod", "ref_mod_ratio",
                              "und_n_reads", "und_prob_mod", "und_mod_ratio")
  
  new_merge_df[is.na(new_merge_df)] <- 0
  
  new_merge_df$n_read_dif = new_merge_df$all_n_reads - (new_merge_df$alt_n_reads + new_merge_df$ref_n_reads + new_merge_df$und_n_reads)

  return(new_merge_df)
}

## unmasked 
max_rep1_all_alt_ref_und_unmasked = make_merge_file(mESC_all_rep1_max, mESC_alt_ALT_rep1_max, mESC_ref_REF_rep1_max, mESC_und_rep1_max)
max_rep2_all_alt_ref_und_unmasked = make_merge_file(mESC_all_rep2_max, mESC_alt_ALT_rep2_max, mESC_ref_REF_rep2_max, mESC_und_rep2_max)

## masked 
max_rep1_all_alt_ref_und = make_merge_file(mESC_all_rep1_max, mESC_alt_masked_rep1_max, mESC_ref_masked_rep1_max, mESC_und_rep1_max)
max_rep2_all_alt_ref_und = make_merge_file(mESC_all_rep2_max, mESC_alt_masked_rep2_max, mESC_ref_masked_rep2_max, mESC_und_rep2_max)

```
## Load vcf data. 
```{r}
## vcf file
mouse_vcf_snp_pre = read.csv("/Users/dayeapark/Desktop/Script_upload/datasets/reference/transcriptomic_variants.vcf", header = FALSE, sep = "\t")

mouse_vcf_snp_pre$gene = sapply(mouse_vcf_snp_pre$V1, rename_default)
mouse_vcf_snp = mouse_vcf_snp_pre[-1, c(6,2,3,4)]
colnames(mouse_vcf_snp) = c("gene", "transcript_pos","ALT", "REF")

```

### probability filter 
```{r}
max_rep1_all_alt_ref_und$Gene = paste0(max_rep1_all_alt_ref_und$gene, "_", max_rep1_all_alt_ref_und$transcript_position)
max_rep2_all_alt_ref_und$Gene = paste0(max_rep2_all_alt_ref_und$gene, "_", max_rep2_all_alt_ref_und$transcript_position)

prob85_start_rep1_masked <- max_rep1_all_alt_ref_und[max_rep1_all_alt_ref_und$all_prob > 0.85, ]
prob85_start_rep2_masked <- max_rep2_all_alt_ref_und[max_rep2_all_alt_ref_und$all_prob > 0.85, ]

prob85_start_rep1_unmasked <- max_rep1_all_alt_ref_und_unmasked[max_rep1_all_alt_ref_und_unmasked$all_prob > 0.85, ]
prob85_start_rep2_unmasked <- max_rep2_all_alt_ref_und_unmasked[max_rep2_all_alt_ref_und_unmasked$all_prob > 0.85, ]



wilcox.test(max_rep1_all_alt_ref_und[max_rep1_all_alt_ref_und$all_prob > 0.85, ]$all_mod_ratio,
            max_mettl3_all_alt_ref_und[max_mettl3_all_alt_ref_und$all_prob > 0.85 ,]$all_mod_ratio, 
            alternative = "two.sided")

```


### filter prob85 first and filter-out DRACH to DRACH sites. 

## Step 1: SNP finder 
### SNP on motif sites were selected from masked all m6Anet output. 
### all, alt, ref, und data frames were merged to all dataset. 
```{r}
SNP_mark <- function(df_nofilter, pos_num, mark) {
  df_pre = df_nofilter[df_nofilter$all_prob > 0.85, ]
  df_pre[is.na(df_pre)] <- 0
  df = df_pre[df_pre$alt_mod_ratio + df_pre$ref_mod_ratio > 0, ]
  
  df$transcript_pos = df$transcript_position + pos_num
  site_mark <- merge(mouse_vcf_snp, df, by = c("gene", "transcript_pos"))
  
  if (nrow(site_mark) > 0 ) { site_mark$SNP_pos = mark}
  
  return(site_mark)
}


SNP_finder_drach <- function(df) { 
    SNP_find_df = rbind(SNP_mark(df, (-1), "D_site"),
                        SNP_mark(df, 0, "R_site"),
                        SNP_mark(df, 1, "A_site"),
                        SNP_mark(df, 2, "C_site"),
                        SNP_mark(df, 3, "H_site") )
    SNP_find_df$Gene = paste0(SNP_find_df$gene, "_", SNP_find_df$transcript_position)
    return(SNP_find_df) 
}

SNP_dependent_m6A_rep1  = SNP_finder_drach(max_rep1_all_alt_ref_und_unmasked)
SNP_dependent_m6A_rep2  = SNP_finder_drach(max_rep2_all_alt_ref_und_unmasked)

SNP_dependent_m6A_rep1_masked  = SNP_finder_drach(max_rep1_all_alt_ref_und)
SNP_dependent_m6A_rep2_masked  = SNP_finder_drach(max_rep2_all_alt_ref_und)
```

## Step 2: DRACH nucleotide check
#In this section, I will identify SNPs that do not shape in the DRACH motif (noDRACH > DRACH or DRACH > noDRACH). 
#Consequently, the identified m6A sites remain unmodified according to the unmasked m6Anet output. I will exclude these cases from consideration.
```{r}
DRACH_both_allele <- function(df) {
  D = c("A", "G", "T")
  R = c("A", "G")
  A = c("A")
  H = c("A", "C", "T")
  
  Dsite = df[df$SNP_pos == "D_site",]
  Rsite = df[df$SNP_pos == "R_site",]
  Asite = df[df$SNP_pos == "A_site",]
  Hsite = df[df$SNP_pos == "H_site",]
  
  Dsite_DRACH = Dsite[Dsite$ALT %in% D & Dsite$REF %in% D, ]
  Rsite_DRACH = Rsite[Rsite$ALT %in% R & Rsite$REF %in% R, ]
  Hsite_DRACH = Hsite[Hsite$ALT %in% H & Hsite$REF %in% H, ]
  DRACH = rbind(Dsite_DRACH, Rsite_DRACH, Hsite_DRACH)
  
  DRACH[is.na(DRACH)] <- 0
  DRACH_select = DRACH[DRACH$alt_mod_ratio + DRACH$ref_mod_ratio > 0,]
  return(DRACH_select)
}

DRACH_prob85_rep1 = DRACH_both_allele(SNP_dependent_m6A_rep1)
DRACH_prob85_rep2 = DRACH_both_allele(SNP_dependent_m6A_rep2)

dim(DRACH_prob85_rep1)
dim(DRACH_prob85_rep2)


```

########## SNP-independent ASM #############
### [PART1] Filter 1 and 2 (remains 40% of raw data)
*Filter 1: transcripts in vcf files 
*Filter 2: alt n_read > 19 & ref n_read > 19
```{r}

# Filter 1 (prob> 0.85 and transcript in vcf)
part1_filter1 <-function (df_nofilter) { 
  df_nofilter[is.na(df_nofilter)] <- 0
  df = df_nofilter[df_nofilter$gene %in% mouse_vcf$gene_name, ]
  
  filter_zero = df_nofilter[(df_nofilter$alt_n_reads + df_nofilter$ref_n_reads) > 0, ]
  
  return(filter_zero)
  }

#part1_rep1_filter1 = part1_filter1(prob85_start_rep1_masked[!(prob85_start_rep1_masked$Gene %in% SNP_dependent_m6A_rep1_masked$Gene), ])
#part1_rep2_filter1 = part1_filter1(prob85_start_rep2_masked[!(prob85_start_rep2_masked$Gene %in% SNP_dependent_m6A_rep2_masked$Gene), ])

##include non-converted DRACH motif sites 
preset_rep1 = rbind (prob85_start_rep1_masked[!(prob85_start_rep1_masked$Gene %in% SNP_dependent_m6A_rep1_masked$Gene), ], 
                     prob85_start_rep1_masked[prob85_start_rep1_masked$Gene %in% DRACH_prob85_rep1$Gene, ])
preset_rep2 = rbind (prob85_start_rep2_masked[!(prob85_start_rep2_masked$Gene %in% SNP_dependent_m6A_rep2_masked$Gene), ], 
                     prob85_start_rep2_masked[prob85_start_rep2_masked$Gene %in% DRACH_prob85_rep2$Gene, ])

part1_rep1_filter1 = part1_filter1(preset_rep1)
part1_rep2_filter1 = part1_filter1(preset_rep2)


## Filter2: all_n_read = (alt_n_read + ref_n_read + und_n_read) or both alt and ref 'n_read' valued higher than 19. 
part1_rep1_filter2 = part1_rep1_filter1[part1_rep1_filter1$ref_n_reads > 19 & part1_rep1_filter1$alt_n_reads > 19, ]
part1_rep2_filter2 = part1_rep2_filter1[part1_rep2_filter1$ref_n_reads > 19 & part1_rep2_filter1$alt_n_reads > 19, ]

## Filter 2 out (for part2)
part1_rep1_filter2_out_mod = part1_rep1_filter1[!(part1_rep1_filter1$ref_n_reads > 19 & part1_rep1_filter1$alt_n_reads > 19), ]
part1_rep2_filter2_out_mod = part1_rep2_filter1[!(part1_rep2_filter1$ref_n_reads > 19 & part1_rep2_filter1$alt_n_reads > 19), ]

```

### [PART2] Re-calcualtion for the other 60%  
* Filter 3: (alt n_read + ref n_read) > 0
* Recalculate mod_ratio 
alt mod ratio =  ((all n_read ∗ all mod ratio) -(ref n_read ∗ ref mod ratio"))/ alt n_read
ref mod ratio =  ((all n_read ∗ all mod ratio) -(alt n_read ∗ alt mod ratio"))/ ref n_read
* Filter 4:  n_read difference > 10
```{r}
## Filter 3: (alt n_read + ref n_read) > 0
part2_rep1_filter2 = part1_rep1_filter2_out_mod[(part1_rep1_filter2_out_mod$alt_n_reads + part1_rep1_filter2_out_mod$ref_n_reads) > 0, ]
part2_rep2_filter2 = part1_rep2_filter2_out_mod[(part1_rep2_filter2_out_mod$alt_n_reads + part1_rep2_filter2_out_mod$ref_n_reads) > 0, ]

## Recount mod ratio 
## 1) alt and ref n_read were filled with n_read differences. 
## 2) if undefined read has modification ratio, I included und_mod_ratio for calculation. However, if und_n_read is zero, I ignore the effects of undefined reads. 
## As a result, recounted mod ratio is not over 1. 

recount_mod_ratio_part2 <- function (df_part2_post_filer2) {
 df_part2_post_filer =  df_part2_post_filer2[df_part2_post_filer2$n_read_dif > 0, ]
 
    for (i in 1:nrow(df_part2_post_filer)) {
      if (df_part2_post_filer$alt_n_reads[i] == 0 ) {
        df_part2_post_filer$alt_n_reads[i] <- df_part2_post_filer$n_read_dif[i]
        
        if (df_part2_post_filer$und_n_reads[i] > 0 ) {
          df_part2_post_filer$alt_mod_ratio[i] = ((df_part2_post_filer$all_n_reads[i] * df_part2_post_filer$all_mod_ratio[i]) - (df_part2_post_filer$ref_n_reads[i] * df_part2_post_filer$ref_mod_ratio[i]) - (df_part2_post_filer$und_n_reads[i] * df_part2_post_filer$und_mod_ratio[i]))/df_part2_post_filer$alt_n_reads[i]  
        } else if (df_part2_post_filer$und_n_reads[i] == 0 )
          df_part2_post_filer$alt_mod_ratio[i] = ((df_part2_post_filer$all_n_reads[i] * df_part2_post_filer$all_mod_ratio[i]) - (df_part2_post_filer$ref_n_reads[i] * df_part2_post_filer$ref_mod_ratio[i]) )/df_part2_post_filer$alt_n_reads[i]
          
      } else if (df_part2_post_filer$ref_n_reads[i] == 0 ) {
        df_part2_post_filer$ref_n_reads[i] <- df_part2_post_filer$n_read_dif[i]
       
        if (df_part2_post_filer$und_n_reads[i] > 0 ) {
         df_part2_post_filer$ref_mod_ratio[i] = ((df_part2_post_filer$all_n_reads[i] * df_part2_post_filer$all_mod_ratio[i]) - (df_part2_post_filer$alt_n_reads[i] * df_part2_post_filer$alt_mod_ratio[i]) - (df_part2_post_filer$und_n_reads[i] * df_part2_post_filer$und_mod_ratio[i] ))/df_part2_post_filer$ref_n_reads[i]
       } else if (df_part2_post_filer$und_n_reads[i] == 0 ) {
         df_part2_post_filer$ref_mod_ratio[i] = ((df_part2_post_filer$all_n_reads[i] * df_part2_post_filer$all_mod_ratio[i]) - (df_part2_post_filer$alt_n_reads[i] * df_part2_post_filer$alt_mod_ratio[i]) )/df_part2_post_filer$ref_n_reads[i]
       }}
    }
      return(df_part2_post_filer)
}

## rep1 and rep2 recount. 
part2_rep1_filter3 = recount_mod_ratio_part2(part2_rep1_filter2)  #  sites 
part2_rep2_filter3 = recount_mod_ratio_part2(part2_rep2_filter2)  #  sites 
## It takes a long time. I recommend you to use the saved file. 

## Filter 4: n_read difference > 10
# I keep m6A sites which n_read differences are over 10
part2_rep1_filter4 = part2_rep1_filter3[part2_rep1_filter3$n_read_dif > 10  , ]
part2_rep2_filter4 = part2_rep2_filter3[part2_rep2_filter3$n_read_dif > 10 , ]

# Remove cases which has undefined mod ratio > 0 
part2_rep1_filter4 = part2_rep1_filter3[part2_rep1_filter3$n_read_dif > 10 & 
                                        part2_rep1_filter3$und_mod_ratio == 0, ]
part2_rep2_filter4 = part2_rep2_filter3[part2_rep2_filter3$n_read_dif > 10 & 
                                        part2_rep2_filter3$und_mod_ratio == 0, ]

```


## Combind part 1 and part 2 for further analysis (motif)
```{r}
## I combined part1 and part2 data tables. 
rep1_part1_part2 = rbind(part1_rep1_filter2, part2_rep1_filter4)
rep2_part1_part2 = rbind(part1_rep2_filter2, part2_rep2_filter4)

step6_prob85 = merge(rep1_part1_part2[,c(1:14)], rep2_part1_part2[,c(1:14)], by = c("gene", "transcript_position") )

colnames(step6_prob85) = c("gene", "transcript_position", 
                           "rep1_all_n_reads", "rep1_all_prob", "rep1_all_mod_ratio",
                           "rep1_alt_n_reads", "rep1_alt_prob", "rep1_alt_mod_ratio",
                           "rep1_ref_n_reads", "rep1_ref_prob", "rep1_ref_mod_ratio",
                           "rep1_und_n_reads", "rep1_und_prob", "rep1_und_mod_ratio",
                           "rep2_all_n_reads", "rep2_all_prob", "rep2_all_mod_ratio",
                           "rep2_alt_n_reads", "rep2_alt_prob", "rep2_alt_mod_ratio",
                           "rep2_ref_n_reads", "rep2_ref_prob", "rep2_ref_mod_ratio",
                           "rep2_und_n_reads", "rep2_und_prob", "rep2_und_mod_ratio")

step6_prob85$Gene = paste0(step6_prob85$gene, "_", step6_prob85$transcript_position)

#write.csv(step6_prob85, file = "/Users/dayeapark/Desktop/Script_upload/datasets/processed_data/step6_prob85.csv", row.names = FALSE, quote = FALSE)
```

## mod differences plot
```{r}
merge_part1_part2 = merge(rep1_part1_part2[,c(1:15)], rep2_part1_part2, by =c("gene", "transcript_position"))

mod_dif_plot <- function ( df ) { 
df$rep1_mod_dif = df$ref_mod_ratio.x - df$alt_mod_ratio.x
df$rep2_mod_dif = df$ref_mod_ratio.y - df$alt_mod_ratio.y

dif_threhold = 0.33

ggplot(df, aes(x= rep1_mod_dif, y = rep2_mod_dif)) + 
  geom_point(size = 1, shape = 19, alpha = 1/2) +
  theme_classic() +
  geom_hline(yintercept=c(-dif_threhold, dif_threhold), linetype="dashed", color = "red", size=0.2) +
  geom_vline(xintercept=c(-dif_threhold, dif_threhold), linetype="dashed", color = "red", size=0.2)

}

mod_dif_plot(merge_part1_part2)

```

############### ASM analysis ###############  
### function for reframe for bootsrap
```{r}
## Detailed SNP information reframe 
reframe_for_bootstrap <- function (m6A_df , prob_threshold) { 
    colnames(m6A_df) = c( "gene", "transcript_position", 
                              "all_n_reads", "all_prob_mod", "all_mod_ratio", 
                              "alt_n_reads", "alt_prob_mod", "alt_mod_ratio",
                              "ref_n_reads", "ref_prob_mod", "ref_mod_ratio",
                              "und_n_reads", "und_prob_mod", "und_mod_ratio")
    
    m6A_allele  = m6A_df[m6A_df$all_prob_mod > prob_threshold,]
    m6A_allele$Gene = paste0(m6A_allele$gene, "_", m6A_allele$transcript_position)
    
    m6A_allele$alt_methylated = m6A_allele$alt_n_reads * m6A_allele$alt_mod_ratio
    m6A_allele$ref_methylated = m6A_allele$ref_n_reads * m6A_allele$ref_mod_ratio
 
    m6A_allele_count = m6A_allele[,c("Gene", "alt_n_reads", "alt_methylated", "alt_mod_ratio","ref_n_reads", "ref_methylated", "ref_mod_ratio")]
    return(m6A_allele_count)
}

```

### function to search for p-value after bootsrapping on m6A site. 
```{r}
set.seed(3)

p_value_of_each_site <- function ( df_all, sample_n, effect_size_threshold) {
  
    df_modcount  = reframe_for_bootstrap(df_all[,c(1:14)], 0.85)
    
    for(i in 1:nrow(df_modcount)) {
      if (df_modcount$alt_n_reads[i] > 0 & df_modcount$ref_n_reads[i] > 0 ) {
        methylated_alt = c(rep(1,df_modcount$alt_methylated[i]), rep(0,df_modcount$alt_n_reads[i] - df_modcount$alt_methylated[i]))
        methylated_ref = c(rep(1,df_modcount$ref_methylated[i]), rep(0,df_modcount$ref_n_reads[i] - df_modcount$ref_methylated[i]))
        
        alt_bootstrap = colSums(replicate(sample(methylated_alt, length(methylated_alt), replace = TRUE), n = sample_n ))/length(methylated_alt)
        ref_bootstrap = colSums(replicate(sample(methylated_ref, length(methylated_ref), replace = TRUE), n = sample_n ))/length(methylated_ref)
        
        pvalue_ref_high = 1- sum((ref_bootstrap - alt_bootstrap) > effect_size_threshold) / length(ref_bootstrap-alt_bootstrap)
        pvalue_alt_high = 1- sum((alt_bootstrap - ref_bootstrap) > effect_size_threshold) / length(ref_bootstrap-alt_bootstrap)
        
        df_modcount$pvalue_ref[i] <- pvalue_ref_high
        df_modcount$pvalue_alt[i] <- pvalue_alt_high
        
      } else if (df_modcount$alt_n_reads[i] == 0 | df_modcount$ref_n_reads[i] == 0 ) {
        print("zero read count found!")
        df_modcount$pvalue_ref[i] = "none"
        df_modcount$pvalue_alt[i] = "none" }
    }
    
return(df_modcount[,c("Gene","alt_mod_ratio","ref_mod_ratio", "pvalue_ref", "pvalue_alt")])
}

## It will take a long time if you use large number for bootstrapping 
## In th paper, we used 10,000 bootstrap. 

mouse_rep1_pvalue = p_value_of_each_site(rep1_part1_part2, 10000, 0.1)
mouse_rep2_pvalue = p_value_of_each_site(rep2_part1_part2, 10000, 0.1)


```

## Select p-value according to the allele-bias
### In this study we obtained one-side p-value where modification differences has positive values. 
### This is the final p-value
```{r}
## pvalue from methylation biasness.
 p_value_bias <- function (df_p_value) { 
    for (i in 1:nrow (df_p_value)) { 
      if(df_p_value$alt_mod_ratio[i] > df_p_value$ref_mod_ratio[i]) { 
        df_p_value$bias_pvalue[i] = df_p_value$pvalue_alt[i]
      } else if(df_p_value$alt_mod_ratio[i] < df_p_value$ref_mod_ratio[i]) { 
        df_p_value$bias_pvalue[i] = df_p_value$pvalue_ref[i]}
      }
   return(df_p_value[,c(1:3,6)]) 
    }
mouse_rep1_biase_p <- p_value_bias(mouse_rep1_pvalue)
mouse_rep2_biase_p <- p_value_bias(mouse_rep2_pvalue)

```

### Final ASM determination
```{r}
mouse_merge_bias_p = merge(mouse_rep1_biase_p %>% select("Gene", contains("pvalue") ),
                           mouse_rep2_biase_p %>% select("Gene", contains("pvalue") ), by = "Gene", all = TRUE)
colnames(mouse_merge_bias_p) = c("Gene", "rep1_pvalue", "rep2_pvalue")
mouse_ASM_pval <- na.omit(mouse_merge_bias_p)

pvalue_cutoff = -log(0.05)
mouse_ASM <- mouse_ASM_pval[-log(mouse_ASM_pval$rep1_pvalue) > pvalue_cutoff & -log(mouse_ASM_pval$rep2_pvalue) > pvalue_cutoff,]

#write.csv(mouse_ASM, file = "/Users/dayeapark/Desktop/Script_upload/datasets/processed_data/mouse_ASM.csv", row.names = FALSE, quote = FALSE)
```

## PLOT: p-value distiribution after boostraps (Fig 3A)
```{r}
pvalue_cutoff = -log(0.05,10)

ggplot(mouse_ASM_pval, aes(x= -log(rep1_pvalue,10), y = -log(rep2_pvalue,10)) )+ 
  geom_point(size = 1, shape = 19, alpha = 1/2) +
  geom_hline(yintercept= pvalue_cutoff, linetype="dashed", color = "red", size=0.2)+
  geom_vline(xintercept= pvalue_cutoff, linetype="dashed", color = "red", size=0.2) + 
  theme_classic() 


```

## PLOT: mod difference, color mark (Figure 3E)
```{r}
mouse_post_filter_merge <- step6_prob85[,c("gene","Gene", "rep1_alt_mod_ratio", "rep1_ref_mod_ratio", "rep2_alt_mod_ratio", "rep2_ref_mod_ratio")]
mouse_post_filter_merge$rep1_mod_dif = mouse_post_filter_merge$rep1_ref_mod_ratio - mouse_post_filter_merge$rep1_alt_mod_ratio
mouse_post_filter_merge$rep2_mod_dif = mouse_post_filter_merge$rep2_ref_mod_ratio - mouse_post_filter_merge$rep2_alt_mod_ratio


### Label ASM 
group1_gene  = unique( c(DRACH_prob85_rep1$Gene, DRACH_prob85_rep2$Gene) )
mouse_post_filter_merge$ASM = "none"
mouse_post_filter_merge[mouse_post_filter_merge$Gene %in% mouse_ASM$Gene, ]$ASM <- "G2_ASM"
mouse_post_filter_merge[mouse_post_filter_merge$Gene %in% mouse_ASM$Gene & 
                        mouse_post_filter_merge$Gene %in% group1_gene, ]$ASM <- "G1_ASM"

dif_threhold = 0.33

## without label 
ggplot(mouse_post_filter_merge, aes(x= rep1_mod_dif, y = rep2_mod_dif, color = ASM)) + 
  geom_point(size = 1.4, shape = 19, alpha = 1/2) +
  scale_color_manual(values = c("none" = "lightgrey", "G2_ASM" = "blue", "G1_ASM" = "magenta") )  + 
  theme_classic()+
  theme(legend.position= "none")

```


## PLOT: alt and ref distribution example (Figure 3B, G1 and G2)
```{r}
set.seed(5)

bootstrap_distance_histogram <- function (df_part1_part2, GOI, sample_n, effect_size_threshold) { 
    df_GOI <- reframe_for_bootstrap(df_part1_part2[df_part1_part2$Gene == GOI, c(1:14)], 0.85)
    
    methylated_alt = c(rep(1,df_GOI$alt_methylated), rep(0,df_GOI$alt_n_reads - df_GOI$alt_methylated))
    methylated_ref = c(rep(1,df_GOI$ref_methylated), rep(0,df_GOI$ref_n_reads - df_GOI$ref_methylated))
            
    alt_bootstrap = colSums(replicate(sample(methylated_alt, length(methylated_alt), replace = TRUE), n = sample_n ))/length(methylated_alt)
    ref_bootstrap = colSums(replicate(sample(methylated_ref, length(methylated_ref), replace = TRUE), n = sample_n ))/length(methylated_ref)
    
    pvalue_ref_high = 1- sum((ref_bootstrap - alt_bootstrap) > effect_size_threshold) / length(ref_bootstrap-alt_bootstrap)
    pvalue_alt_high = 1- sum((alt_bootstrap - ref_bootstrap) > effect_size_threshold) / length(ref_bootstrap-alt_bootstrap)
    
    plot_df <- data.frame( alt_bootstrap_pop = alt_bootstrap, ref_bootstrap_pop = ref_bootstrap)
    
    hist <- ggplot(melt(plot_df), aes(x=value, fill=variable)) +
          geom_histogram(position="identity", bins=20, alpha = 0.6) + 
          theme_classic() +
          theme(legend.position= "none")
    
    print(paste("pvalue ref: ", pvalue_ref_high, "pvalue alt: ",pvalue_alt_high))
    
    return(hist) 
}


ASM_Trim25 <- bootstrap_distance_histogram(rep1_part1_part2, "Trim25_4580", 10000, 0.1)
nonASM_Trim59 <- bootstrap_distance_histogram(rep1_part1_part2, "Trim59_1195", 10000, 0.1)

ggarrange(ASM_Trim25, nonASM_Trim59, nrow =1)


```


## PLOT: heatmap of mod ratio (Figure 3C)
```{r}
mouse_ASM_mod <- step6_prob85[step6_prob85$Gene %in% mouse_ASM$Gene, ]

library(pheatmap)

ASM_heatmap_df = mouse_ASM_mod[, c("Gene", "rep1_alt_mod_ratio", "rep1_ref_mod_ratio",
                                   "rep2_alt_mod_ratio", "rep2_ref_mod_ratio")]
rownames(ASM_heatmap_df) = ASM_heatmap_df$Gene

## heatmap from ggplot 
Name_order = c("Armc10_1170", "Dnpep_1287", "Glmp_1308", "Nudt1_639", "L3mbtl2_3081", 
          "Tmbim6_2019", "Gcsh_1211", "Trim25_4580", "Kif11_4336", "Atp5o_777", 
          "Psrc1_1465", "Pml_2377", "Armc10_1624", "Gnpat_1888", "Rsl1d1_1773", 
          "Ccnb1_1470", "Ddx20_2200", "Kif5b_4766", "Cmtm7_591", "2810004N23Rik_1346", 
          "Uhrf1_3221", "Atmin_4453", "Dgcr2_3682", "Stk38_2863")

name_ordered <- ASM_heatmap_df %>% mutate(name = fct_relevel(Gene, Name_order) )
                            
ASM_heatmap_ggplto_df <- melt(name_ordered, id = c("Gene", "name") )


ggplot(ASM_heatmap_ggplto_df, aes(variable, fct_reorder(name,Gene), fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0.5, limit = c(0,1), space = "Lab", 
    name="mod ratio") + 
  theme_classic()


```
